students/franjaku/Lesson_5/mailroom_3.py

```diff
@@ -0,0 +1,181 @@
+    amount = input(donation_amount_prompt)
+
+    try:
+        amount = float(amount)
+    except ValueError:
+        print('Not a valid number.')
+        return get_donation_amount()
+
+    return amount
+
+
+def thank_you_note_prompt():
+    prompt = 'Enter the donors full name or type "list" to see all donors in the database.\n>>> '
+    donor_name = (input(prompt)).title()
+
+    if donor_name == '':
+        donor_name = 'List'
+
+    return donor_name
+
+
+def send_thank_you_note(database):
+
+    # Compose an email: thank the user, print email to terminal
+    Email = '{} thank you for your generous donation of ${:.2f}'
+
+    donor_name = thank_you_note_prompt()
+
+    if donor_name == 'List':
+        print_donor_list(database)
+    elif donor_name in database:
+        donation_amount = get_donation_amount()
+        database = add_new_donation(database, donor_name, donation_amount)
+        print(Email.format(donor_name, donation_amount))
+    elif donor_name not in database:
+        donation_amount = get_donation_amount()
+        database = add_new_donor(database, donor_name, donation_amount)
+        print(Email.format(donor_name, donation_amount))
+    else:
+        # function error catching
+        print('somethings not right')
+
+    return None
+
+
+def send_thank_you_note_all(database):
+
+    for donor, data in database.items():
+        create_letter(donor, data[2])
+
+
+def create_letter(name, total_donation):
+    with open(f"{name}.txt", 'w+') as outfile:
+        outfile.write(f"Dear {name}\n")
+        outfile.write(f"Thank you for you donation of ${total_donation}!\n")
+        outfile.write(f"Best wishes,\nThe Team")
+    outfile.close()
+
+
+def create_report(database):
+    # Sort database by most $$ donated
+    sorted_database = collections.OrderedDict(sorted(database.items(), key=operator.itemgetter(1), reverse=True))
+    print('-----Donation Report-----')
+    print('\n{:<15} | {:>14} | {:>11} | {:>16}'.format('Donor Name', 'Total Donation', '# donations', 'Average Donation'))
+    print('-'*(15+14+11+16+10))
+    for donor, data in sorted_database.items():
+        print(f"{donor:<15} | ${data[0]:>13.2f} | {data[1]:^11} | ${data[2]:>15.2f}")
+
+    print('\n')
+
+
+def quit(database=None):
+    print('Goodbye!')
+    os.sys.exit()
+
+
+# Driver Function
+def mail_room():
+    database = initialize_database()
+    welcome_message = '------------Welcome to the Mailroom :)------------'
+    print(welcome_message)
+
+    options_dict = {
+        '1': send_thank_you_note,
+        '2': create_report,
+        '3': send_thank_you_note_all,
+        '4': quit}
+
+    while True:
+        UserAction = prompt_user()
+
+        if UserAction in options_dict:
+            options_dict.get(UserAction)(database)
+        else:
+            print('Not a valid option...\n')
+
+    return None
+
+
+# Main Interaction
+if __name__ == '__main__':
+    mail_room()
```

Comment:
@franjaku - not seeing comprehensions or exceptions in this version of mailroom.  Is this complete and ready for review?  Please take a look again at the requirements for this assignment and let me know if you have questions.